File: exercism/haskell/space-age/src/SpaceAge.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module SpaceAge
  (Planet(..)
  ,ageOn)
  where

newtype Seconds = S Double deriving (Eq, Ord, Num, Fractional, Real)

data Planet
    = Mercury
    | Venus
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune

ageOn :: Planet -> Seconds -> Seconds
ageOn p = (/ (period * earthYear))
 where
  earthYear  = 31557600
  period = case p of
    Earth   -> 1
    Mercury -> 0.2408467
    Venus   -> 0.61519726
    Mars    -> 1.8808158
    Jupiter -> 11.862615
    Saturn  -> 29.447498
    Uranus  -> 84.016846
    Neptune -> 164.79132

File: exercism/ocaml/space-age/space_age.ml
type planet =
  | Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Neptune
  | Uranus

let age_on p s =
  let m =
    match p with
    | Mercury -> 0.2408467
    | Venus -> 0.61519726
    | Earth -> 1.0
    | Mars -> 1.8808158
    | Jupiter -> 11.862615
    | Saturn -> 29.447498
    | Uranus -> 84.016846
    | Neptune -> 164.79132
  in
  float_of_int s /. (m *. 31557600.0)
